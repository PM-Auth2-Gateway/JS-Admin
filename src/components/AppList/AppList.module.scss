.listItem:hover {
  background-color: rgba(239, 240, 242, 0.5);
}
