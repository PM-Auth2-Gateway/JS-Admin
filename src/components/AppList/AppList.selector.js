import { createSelector } from 'reselect';

const selector = createSelector(
  (globalState) => globalState.apps,
  (apps) => apps.all
);

export default selector;
