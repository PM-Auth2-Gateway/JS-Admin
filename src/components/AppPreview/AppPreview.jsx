import React from 'react';
import { Link } from 'react-router-dom';

import styles from './AppPreview.module.scss';

import appIcon from '../../assets/app-icon.svg';

const AppPreview = ({ url, name }) => {
  return (
    <div className={styles.appPreview}>
      <Link to={url}>
        <img src={appIcon} className={styles.appIcon} alt='icon' />
      </Link>
      <span className={styles.appName}>
        <Link to={url}>{name}</Link>
      </span>
    </div>
  );
};

export default AppPreview;
