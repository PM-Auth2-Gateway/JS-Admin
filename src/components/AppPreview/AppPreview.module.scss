.appPreview {
  display: flex;
  align-items: flex-start;
  gap: 16px;

  .appIcon {
    width: 40px;
    border-radius: 4px;
    background-color: #eff0f2;
  }

  .appName {
    font-size: 18px;

    a {
      color: #0A84AE;
      text-decoration: none;

      &:hover {
        color: #053B4E;
      }
    }
  }
}
