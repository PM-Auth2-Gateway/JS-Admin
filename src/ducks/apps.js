import config from '../config';
import AppsApiService from '../services/AppsApiService';

const LOAD_ALL = 'apps/all/load';
const ADD_ALL = 'apps/all/add';
const DELETE_BY_ID_ALL = 'apps/all/delete_by_id';
const CLEAR_ALL = 'apps/all/clear';

const SET_CURRENT = 'apps/current/set';
const REMOVE_CURRENT = 'apps/current/delete';

const SET_ERROR = 'apps/error/set';

export const loadAllApps = () => (dispatch) => {
  AppsApiService.get(config.authToken)
    .then((response) => {
      dispatch(loadAll(response.data));
    })
    .catch((error) => {
      dispatch(setError(error));
    });
};

export const addApp = (name) => (dispatch) => {
  debugger;
  AppsApiService.post(config.authToken, { name })
    .then((response) => {
      dispatch(addAll(response.data));
    })
    .catch((error) => {
      dispatch(setError(error));
    });
};

export const deleteAppById = (id) => (dispatch) => {
  AppsApiService.deleteById(config.authToken, { id })
    .then((response) => {
      dispatch(deleteByIdAll(id));
    })
    .catch((error) => {
      dispatch(setError(error));
    });
};

export const clearAllApps = () => (dispatch) => {
  dispatch(clearAll());
};

export const loadCurrentApp = (id) => (dispatch) => {
  AppsApiService.getById(config.authToken, { id })
    .then((response) => {
      dispatch(setCurrent(response.data));
    })
    .catch((error) => {
      dispatch(setError(error));
    });
};

export const removeCurrentApp = () => (dispatch) => {
  dispatch(removeCurrent());
};

export const setAppError = (msg) => (dispatch) => {
  dispatch(setError(msg));
};

const loadAll = (apps) => ({
  type: LOAD_ALL,
  payload: apps,
});

const addAll = (app) => ({
  type: ADD_ALL,
  payload: app,
});

const deleteByIdAll = (id) => ({
  type: DELETE_BY_ID_ALL,
  payload: id,
});

const clearAll = () => ({
  type: CLEAR_ALL,
});

const setCurrent = (app) => ({
  type: SET_CURRENT,
  payload: app,
});

const removeCurrent = () => ({
  type: REMOVE_CURRENT,
});

const setError = (msg) => ({
  type: SET_CURRENT,
  payload: msg,
});

const initialState = {
  all: [],
  current: {},
  error: {},
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case LOAD_ALL:
      return {
        ...state,
        all: [...state.all, ...action.payload],
      };
    case ADD_ALL:
      return {
        ...state,
        all: [...state.all, action.payload],
      };
    case DELETE_BY_ID_ALL:
      return {
        ...state,
        all: state.all.filter((app) => app.id !== action.payload.id),
      };
    case CLEAR_ALL:
      return {
        ...state,
        all: [],
      };
    case SET_CURRENT:
      return {
        ...state,
        current: action.payload,
      };
    case REMOVE_CURRENT:
      return {
        ...state,
        current: null,
      };
    case SET_ERROR:
      return {
        ...state,
        error: action.payload,
      };
    default:
      return state;
  }
};
