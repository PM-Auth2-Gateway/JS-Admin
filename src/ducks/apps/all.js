import AppsApiService from '../../services/AppsApiService';

const LOAD_ALL = 'apps/all/load';
const ADD_ALL = 'apps/all/add';
const CLEAR_ALL = 'apps/all/clear';

const SET_LOADING = 'apps/all/loading/set';
const SET_ERROR = 'apps/all/error/set';

export const loadAllApps = () => async (dispatch) => {
  try {
    dispatch(setLoading(true));
    const res = await AppsApiService.get();

    dispatch(loadAll(res.data));
    dispatch(setError(null));
  } catch (error) {
    dispatch(setError(error));
  } finally {
    dispatch(setLoading(false));
  }
};

export const addApp = ({ name }) => async (dispatch) => {
  try {
    dispatch(setLoading(true));
    const res = await AppsApiService.post({ name });

    dispatch(addAll(res.data));
    dispatch(setError(null));
  } catch (error) {
    dispatch(setError(error));
  } finally {
    dispatch(setLoading(false));
  }
};

export const clearAllApps = () => (dispatch) => {
  dispatch(clearAll());
};

export const setAppError = (msg) => (dispatch) => {
  dispatch(setError(msg));
};

const loadAll = (apps) => ({
  type: LOAD_ALL,
  payload: apps,
});

const addAll = (app) => ({
  type: ADD_ALL,
  payload: app,
});

const clearAll = () => ({
  type: CLEAR_ALL,
});

const setLoading = (status) => ({
  type: SET_LOADING,
  payload: status,
});

const setError = (msg) => ({
  type: SET_ERROR,
  payload: msg,
});

const initialState = {
  value: [],
  loading: false,
  error: null,
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case LOAD_ALL:
      return {
        ...state,
        value: [...state.value, ...action.payload],
      };
    case ADD_ALL:
      return {
        ...state,
        value: [...state.value, action.payload],
      };
    case CLEAR_ALL:
      return {
        ...state,
        value: [],
      };
    case SET_LOADING:
      return {
        ...state,
        loading: action.payload,
      };
    case SET_ERROR:
      return {
        ...state,
        error: action.payload,
      };
    default:
      return state;
  }
};
