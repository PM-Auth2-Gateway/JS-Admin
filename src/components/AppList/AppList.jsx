import PropTypes from 'prop-types';
import { ListGroup, Spinner } from 'react-bootstrap';

import styles from './AppList.module.scss';

import AppPreview from '../AppPreview/AppPreview';

const propTypes = {
  list: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.object,
};

const AppList = (props) => {
  const { list, loading } = props;

  return loading ? (
    <Spinner animation='border' />
  ) : (
    <ListGroup variant={'flush'}>
      {list.map(({ id, name }) => (
        <ListGroup.Item key={id} className={styles.listItem}>
          <AppPreview url={`applications/${id}`} name={name} />
        </ListGroup.Item>
      ))}
    </ListGroup>
  );
};

AppList.propTypes = propTypes;

export default AppList;
