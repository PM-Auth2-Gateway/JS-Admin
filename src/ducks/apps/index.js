import { combineReducers } from 'redux';

import { reducer as appsAllReducer } from './all';
import { reducer as appsCurrentReducer } from './current';

export const reducer = combineReducers({
  all: appsAllReducer,
  current: appsCurrentReducer,
});
