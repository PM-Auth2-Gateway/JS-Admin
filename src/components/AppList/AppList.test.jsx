import React from 'react';
import { shallow } from 'enzyme';
import { useRouteMatch, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { ListGroup } from 'react-bootstrap';

import AppList from './AppList';
import AppPreview from '../AppPreview/AppPreview';

import { clearAllApps, loadAllApps } from '../../ducks/apps/all';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: jest.fn(),
  useHistory: jest.fn(),
  Link: () => '<div />',
}));

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../../ducks/apps/all', () => ({
  clearAllApps: jest.fn(),
  loadAllApps: jest.fn(),
}));

describe('AppList Component', () => {
  const dispatch = jest.fn();
  const action = jest.fn();

  const historyPush = jest.fn();

  beforeAll(() => {
    useRouteMatch.mockReturnValue({ url: '/applications/1' });

    clearAllApps.mockReturnValue(action);
    loadAllApps.mockReturnValue(action);
  });

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);

    useHistory.mockReturnValue({
      push: historyPush,
    });
    useRouteMatch.mockReturnValue({ url: '/applications/' });
    useSelector.mockReturnValue({
      all: [
        {
          id: 1,
          name: 'Test1',
        },
        {
          id: 2,
          name: 'Test2',
        },
      ],
      loading: false,
      error: null,
    });
  });

  it('should render AppList Component', () => {
    const component = shallow(<AppList />);

    expect(component).toMatchSnapshot();
  });

  it('should render to AppPreview', () => {
    const component = shallow(<AppList />);

    const list = component.find(ListGroup);

    const apps = list.find(AppPreview);

    expect(apps.length).toBe(2);
  });
});
