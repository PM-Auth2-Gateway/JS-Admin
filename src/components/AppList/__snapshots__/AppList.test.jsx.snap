// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppList Component should render AppList Component 1`] = `
<Container
  fluid={false}
>
  <Row
    className="mb-5 align-items-center"
    noGutters={false}
  >
    <Col>
      <h2>
        Applications
      </h2>
    </Col>
    <Col
      xs="auto"
    >
      <CreateAppModal />
    </Col>
  </Row>
  <ListGroup
    variant="flush"
  >
    <ListGroupItem
      active={false}
      className="listItem"
      disabled={false}
      key="1"
    >
      <AppPreview
        name="Test1"
        url="/applications//1"
      />
    </ListGroupItem>
    <ListGroupItem
      active={false}
      className="listItem"
      disabled={false}
      key="2"
    >
      <AppPreview
        name="Test2"
        url="/applications//2"
      />
    </ListGroupItem>
  </ListGroup>
</Container>
`;
